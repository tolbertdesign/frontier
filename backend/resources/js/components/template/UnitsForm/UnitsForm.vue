<template>
  <div>
    <form
      v-if="canEdit && unitRange.length > 0"
      class="select"
      @submit.prevent=""
    >
      <select
        v-model="newUnits"
        name="units"
        class="font-bold units-containers"
        @change="updateUnits()"
      >
        <option
          v-for="index in unitRange"
          :key="index"
          :value="index"
        >
          {{ index }}
        </option>
      </select>
    </form>
    <p
      v-else
      class="px-5 border units-container border-grey-light border-solid rounded sm:py-2 py-1 font-bold"
    >
      {{ newUnits }}
    </p>
  </div>
</template>

<script>
import SaveSuccessModal from '@/components/template/SaveSuccessModal'
import Blur from '@/mixins/Blur'

export default {
  name: 'UnitsForm',
  status: 'prototype',
  release: '1.0.0',
  mixins: [Blur],
  props: {
    participantUserId: {
      type: Number,
      default: null,
    },
    units: {
      type: Number,
      default: 0,
    },
    unitNamePlural: {
      type: String,
      default: 'Laps',
    },
    unitMax: {
      type: Number,
      default: 35,
    },
    canEdit: {
      type: Boolean,
      default: true,
    },
  },
  data () {
    return {
      unitRangeLow: 10,
      newUnits: null,
    }
  },
  computed: {
    lang () {
      return this.$store.state.lang
    },
    successModalTitle () {
      return this.parseLanguage(this.lang.update_units_success_title, { name_plural: this.unitNamePlural })
    },
    successModalSubtitle () {
      return this.parseLanguage(this.lang.update_units_success_subtitle, { name_plural: this.unitNamePlural })
    },
    unitRange () {
      const range = []
      for (let i = this.unitRangeLow; i <= this.unitMax; i++) {
        range.push(i)
      }
      return range
    },
  },
  mounted () {
    if (this.units > 0) {
      this.newUnits = this.units
    }
  },
  methods: {
    updateUnits () {
      if (this.newUnits > 0 && this.newUnits <= this.unitMax) {
        this.$axon.update('users', this.participantUserId, {
          units: this.newUnits,
        }).then(response => {
          this.$buefy.modal.open({
            parent: this,
            component: SaveSuccessModal,
            hasModalCard: true,
            props: {
              title: this.successModalTitle,
              subtitle: this.successModalSubtitle,
              duration: 0,
            },
            canCancel: ['escape', 'outside'],
            onCancel: this.unBlur,
            width: 'auto',
          })
        }).catch(error => {
          console.log(error)
        })
      }
    },
  },
}
</script>
