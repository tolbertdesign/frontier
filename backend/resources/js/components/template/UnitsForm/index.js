import UnitsForm from './UnitsForm'

export default UnitsForm
