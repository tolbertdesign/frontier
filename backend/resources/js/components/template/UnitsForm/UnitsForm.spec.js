import { shallowMount, createLocalVue } from '@vue/test-utils'
import UnitsForm from './UnitsForm.vue'

const localVue = createLocalVue()

describe('UnitsForm', () => {
  it('renders units form correctly when editable', () => {
    const wrapper = shallowMount(UnitsForm, {
      propsData: {
        units: 30,
        participantUserId: 5,
        unitNamePlural: 'laps',
        unitMax: 35,
      },
      localVue,
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('renders units form correctly when not editable', () => {
    const wrapper = shallowMount(UnitsForm, {
      propsData: {
        units: 30,
        participantUserId: 6,
        unitNamePlural: 'reading challenges',
        unitMax: 35,
        canEdit: false,
      },
      localVue,
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('displays the correct number of options in the unit select input', () => {
    const wrapper = shallowMount(UnitsForm, {
      propsData: {
        units: 15,
        participantUserId: 7,
        unitNamePlural: 'laps',
        unitMax: 19,
        canEdit: true,
      },
      localVue,
    })
    expect(wrapper.vm.unitRange.length).toEqual(10)
  })
})
