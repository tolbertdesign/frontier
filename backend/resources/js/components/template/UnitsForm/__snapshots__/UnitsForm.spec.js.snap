// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitsForm renders units form correctly when editable 1`] = `
<div>
  <form class="select"><select name="units" class="font-bold units-containers">
      <option value="10">
        10
      </option>
      <option value="11">
        11
      </option>
      <option value="12">
        12
      </option>
      <option value="13">
        13
      </option>
      <option value="14">
        14
      </option>
      <option value="15">
        15
      </option>
      <option value="16">
        16
      </option>
      <option value="17">
        17
      </option>
      <option value="18">
        18
      </option>
      <option value="19">
        19
      </option>
      <option value="20">
        20
      </option>
      <option value="21">
        21
      </option>
      <option value="22">
        22
      </option>
      <option value="23">
        23
      </option>
      <option value="24">
        24
      </option>
      <option value="25">
        25
      </option>
      <option value="26">
        26
      </option>
      <option value="27">
        27
      </option>
      <option value="28">
        28
      </option>
      <option value="29">
        29
      </option>
      <option value="30">
        30
      </option>
      <option value="31">
        31
      </option>
      <option value="32">
        32
      </option>
      <option value="33">
        33
      </option>
      <option value="34">
        34
      </option>
      <option value="35">
        35
      </option>
    </select></form>
</div>
`;

exports[`UnitsForm renders units form correctly when not editable 1`] = `
<div>
  <p class="px-5 border units-container border-grey-light border-solid rounded sm:py-2 py-1 font-bold">
    30
  </p>
</div>
`;
